### DISCUSSION TITLE
FR: A way to get the siteId currently being processed in the EVENT_REGISTER_SEARCHABLE_ATTRIBUTES event

### DISCUSSION THREAD

**Discussion by rob-c-baker** [craftcms/cms#8852] [Ideas]
### Description

The `RegisterElementSearchableAttributesEvent` class only has access to the array of attributes / field handles to be indexed. I need a way to change the array I return from that event based on which site is currently having it's search index built.

The use case is that I have a custom field that allows me to enable / disable categories (and certain specific matrix blocks amongst a  whole host of other things) on specific sites on the front end, which works great, except the search index gets built with keywords that no longer appear on the pages / entries they reference.

The reason for the custom field on the categories is that it is not possible to enable / disable categories on specific sites as it is with entries as the time of writing.

If there is another way to do what I need, that would be awesome. Otherwise, a `siteId` in that event class would work a treat.

**Comment by brandonkelly** [comment]
This isn’t currently possible as `EVENT_REGISTER_SEARCHABLE_ATTRIBUTES` (and elements’ `searchableAttributes()` method) is implemented at the static level, so the attributes are defined outside of any specific instance of the element.

We should be able to clean this up for Craft 4 though, perhaps by removing elements’ static `searchableAttributes()` method altogether, and removing the `$attribute` argument from `getSearchKeywords()`, and have elements reimplement `getSearchKeywords()` to return an array of attribute-keyword pairs.

**Comment by rob-c-baker** [comment]
Thanks @brandonkelly that sounds like a plan.

I heard somewhere that categories (and tags) might going in favour of a more generic content element type in Craft 4. That being the case, it would probably remove my use case, but I can see how being able to access the site and potentially other instance properties from that event could be useful.

Now I just need to work out a way of working around my workaround and get the search indexes in a more accurate state!
